src/prometheus/prometheus.c

```diff
@@ -0,0 +1,358 @@
+  }
+
+  if ((verify_host == NULL) || IS_TRUE(verify_host)) {
+    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 2L);
+  } else {
+    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
+  }
+
+  if (cacert != NULL) {
+    curl_easy_setopt(curl, CURLOPT_CAINFO, cacert);
+  }
+
+#ifdef HAVE_CURLOPT_TIMEOUT_MS
+  if (timeout != NULL) {
+    curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS, atol(timeout));
+  } else {
+    curl_easy_setopt(curl, CURLOPT_TIMEOUT_MS,
+                     (long)CDTIME_T_TO_MS(plugin_get_interval()));
+  }
+#endif
+
+#ifdef HAVE_CURLOPT_UNIX_SOCKET_PATH
+  if (sock != NULL) {
+    curl_easy_setopt(curl, CURLOPT_UNIX_SOCKET_PATH, sock);
+  }
+#endif
+
+  static char jwt_header[512];
+  int status = ssnprintf(jwt_header, sizeof(jwt_header),
+                         "Authorization: Bearer %s", jwt_token);
+  if ((status < 0) || ((size_t)status >= sizeof(jwt_header))) {
+    ERROR("prometheus plugin: jwt token would have been truncated.");
+    return -1;
+  }
+  curl_easy_setopt(curl, CURLOPT_HTTPHEADER,
+                   curl_slist_append(NULL, jwt_header));
+
+  return 0;
+}
+
+static metric_t *convert_pr_metric_to_metric(pr_metric_t *pr_metric,
+                                             metric_type_t metric_type) {
+  metric_t *metric = malloc(sizeof(*metric));
+  if (!metric) {
+    ERROR("Couldn't allocate memory for metric");
+    return NULL;
+  }
+  memset(metric, 0, sizeof(*metric));
+  pr_label_t *cur_pr_label = pr_metric->labels;
+  while (cur_pr_label) {
+    if (metric_label_set(metric, cur_pr_label->name, cur_pr_label->value) !=
+        0) {
+      metric_reset(metric);
+      free(metric);
+      return NULL;
+    }
+    cur_pr_label = cur_pr_label->next;
+  }
+  switch (metric_type) {
+  case (METRIC_TYPE_UNTYPED): {
+    metric->value.counter_fp = pr_metric->value;
+    break;
+  }
+  case (METRIC_TYPE_COUNTER_FP): {
+    metric->value.counter_fp = pr_metric->value;
+    break;
+  }
+  case (METRIC_TYPE_UP_DOWN_FP): {
+    metric->value.up_down_fp = pr_metric->value;
+    break;
+  }
+  default: {
+    ERROR("Unsupported prometheus value representation");
+    metric_reset(metric);
+    free(metric);
+    return NULL;
+  }
+  }
+  if (pr_metric->timestamp->has_value) {
+    metric->time = pr_metric->timestamp->value;
+  }
+  return metric;
+}
+
+static int convert_pr_fam_to_fam(pr_metric_family_t *pr_fam,
+                                 metric_family_t *fam) {
+  switch (pr_fam->tp) {
+  case (PR_COUNTER): {
+    fam->type = METRIC_TYPE_COUNTER_FP;
+    break;
+  }
+  case (PR_GAUGE): {
+    fam->type = METRIC_TYPE_UP_DOWN_FP;
+    break;
+  }
+  case (PR_UNTYPED): {
+    fam->type = METRIC_TYPE_UNTYPED;
+    break;
+  }
+  case (PR_SUMMARY): {
+    fam->type = METRIC_TYPE_UP_DOWN_FP;
+    break;
+  }
+  case (PR_HISTOGRAM): {
+    fam->type = METRIC_TYPE_UP_DOWN_FP;
+    break;
+  }
+  default: {
+    ERROR("Unknown prometheus metric family type");
```

Comment:
If there's an error, log message should give details on the problem, i.e. what the unknown received type was.

(And please properly capitalize Prometheus name.)